test_api.go

```diff
@@ -0,0 +1,327 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/bitrise-io/go-utils/log"
+	testing "google.golang.org/api/testing/v1"
+)
+
+// TestAsset describes a requested test asset
+type TestAsset struct {
+	UploadURL string `json:"uploadUrl"`
+	GcsPath   string `json:"gcsPath"`
+	Filename  string `json:"filename"`
+}
+
+// TestAssetsAndroid describes requested Android test asset and as the returned test asset upload URLs
+type TestAssetsAndroid struct {
+	isBundle   bool
+	testApp    *TestAsset
+	Apk        TestAsset   `json:"apk,omitempty"`
+	Aab        TestAsset   `json:"aab,omitmepty"`
+	TestApk    TestAsset   `json:"testApk,omitempty"`
+	RoboScript TestAsset   `json:"roboScript,omitempty"`
+	ObbFiles   []TestAsset `json:"obbFiles,omitempty"`
+}
+
+func uploadTestAssets(configs ConfigsModel) (TestAssetsAndroid, error) {
+	var testAssets TestAssetsAndroid
+
+	url := configs.APIBaseURL + "/assets/android/" + configs.AppSlug + "/" + configs.BuildSlug + "/" + configs.APIToken
+
+	if strings.ToLower(filepath.Ext(configs.AppPath)) == ".aab" {
+		testAssets.isBundle = true
+	}
+	log.Debugf("App path (%s), is bundle: %t", configs.AppPath, testAssets.isBundle)
+
+	var requestedAssets TestAssetsAndroid
+	if testAssets.isBundle {
+		requestedAssets.Aab = TestAsset{
+			Filename: filepath.Base(configs.AppPath),
+		}
+	} else {
+		requestedAssets.Apk = TestAsset{
+			Filename: filepath.Base(configs.AppPath),
+		}
+	}
+	if configs.TestType == testTypeInstrumentation {
+		requestedAssets.TestApk = TestAsset{
+			Filename: filepath.Base(configs.TestApkPath),
+		}
+	}
+	if configs.TestType == testTypeRobo && configs.RoboScenarioFile != "" {
+		requestedAssets.RoboScript = TestAsset{
+			Filename: filepath.Base(configs.RoboScenarioFile),
+		}
+	}
+	for _, obbFile := range configs.ObbFiles {
+		requestedAssets.ObbFiles = append(requestedAssets.ObbFiles, TestAsset{
+			Filename: filepath.Base(obbFile),
+		})
+	}
+
+	log.Debugf("Assets requested: %+v", requestedAssets)
+
+	data, err := json.Marshal(requestedAssets)
+	if err != nil {
+		return TestAssetsAndroid{}, fmt.Errorf("failed to encode to json: %+v", requestedAssets)
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewReader(data))
+	if err != nil {
+		return TestAssetsAndroid{}, fmt.Errorf("failed to create http request, error: %s", err)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return TestAssetsAndroid{}, fmt.Errorf("failed to get http response, error: %s", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		body, err := ioutil.ReadAll(resp.Body)
```

Comment:
These lines can be moved out from if and you could evade code duplication. (Line 91-94 and 98-101)

Reply:
fixed